SQL

**Цель работы** — проанализировать базу данных крупного сервиса для чтения книг по подписке, содержащую информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

Для этого необходимо решить следующие задачи:

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


<a id = 'section1'></a>
# Оглавление проекта

<a class="ancor" id="step_0"></a>
* [Исследование таблиц](#step_1)
* [ЗАДАНИЯ](#step_2)
 * [1  Посчитайте, сколько книг вышло после 1 января 2000 года](#step_3)
 * [2  Для каждой книги посчитайте количество обзоров и среднюю оценку](#step_4)
 * [3  Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры](#step_5)
 * [4  Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками](#step_6)
 * [5  Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#step_7)

<a class="ancor" id="step_1"></a>
# Исследование таблиц

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
            'pwd': 'Sdf4$2;d-d30pp', # пароль
            'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
            'port': 6432, # порт подключения
            'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                db_config['pwd'],
                                                db_config['host'],
                                                db_config['port'],
                                                db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [3]:
!pip3 install psycopg2-binary

Defaulting to user installation because normal site-packages is not writeable


<a href = '#section1'>Вернуться к оглавлению</a>
</div>

In [4]:
books = ''' SELECT
                  * 
                FROM books
            LIMIT 5;   
            '''
pd.io.sql.read_sql(books, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Содержит данные о книгах:                                                                                         
book_id — идентификатор книги;                                                                                     
author_id — идентификатор автора;                                                                                 
title — название книги;                                                                                           
num_pages — количество страниц;                                                                                   
publication_date — дата публикации книги;                                                                         
publisher_id — идентификатор издателя. 

In [5]:
authors = ''' SELECT * 
            FROM authors
            LIMIT 5;  
            '''
pd.io.sql.read_sql(authors, con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Содержит данные об авторах:                                                                                       
author_id — идентификатор автора;                                                                                 
author — имя автора.                                                                                               

In [6]:
publishers = ''' SELECT * 
            FROM publishers
            LIMIT 5;  
            '''
pd.io.sql.read_sql(publishers, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Содержит данные об издательствах:                                                                                         
publisher_id — идентификатор издательства;                                                                                
publisher — название издательства.

In [7]:
ratings = ''' SELECT * 
            FROM ratings
            LIMIT 5;  
            '''
pd.io.sql.read_sql(ratings, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Содержит данные о пользовательских оценках книг:                                                                           
rating_id — идентификатор оценки;                                                                                         
book_id — идентификатор книги;                                                                                             
username — имя пользователя, оставившего оценку;                                                                          
rating — оценка книги.

In [8]:
reviews = ''' SELECT * 
            FROM reviews
            LIMIT 5;  
            '''
pd.io.sql.read_sql(reviews, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Содержит данные о пользовательских обзорах на книги:                                                               
review_id — идентификатор обзора;                                                                                
book_id — идентификатор книги;                                                                                    
username — имя пользователя, написавшего обзор;                                                                   
text — текст обзора.

ВЫВОД:  

Выведены первые строки  таблиц и изучены данные в них

<a href = '#section1'>Вернуться к оглавлению</a>
</div>

<a class="ancor" id="step_2"></a>
# ЗАДАНИЯ

<a class="ancor" id="step_3"></a>
## Посчитайте, сколько книг вышло после 1 января 2000 года

In [9]:
query_1 = ''' SELECT 
                  COUNT (DISTINCT book_id) AS count
              FROM 
                  books 
              WHERE 
                  publication_date >= '2000-01-01';
          '''
pd.io.sql.read_sql(query_1, con = engine)

count
0    821

ВЫВОД:

После 1 января 2000 вышлa 821 книга


<a href = '#section1'>Вернуться к оглавлению</a>
</div>

<a class="ancor" id="step_4"></a>
## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [10]:
query_2 = '''SELECT
                  rat.book_id AS book_id,
                  rat.title AS title,
                  rat.author AS author,
                  rat.rating AS rating,
                  rev.reviews_count AS reviews_count
             FROM
            (SELECT
                  books.book_id AS book_id,
                  books.title AS title,
                  authors.author AS author,
                  AVG(ratings.rating) AS rating
              FROM
                  books
                  LEFT JOIN authors ON authors.author_id = books.author_id
                  LEFT JOIN ratings ON ratings.book_id = books.book_id
              GROUP BY
                  books.book_id,
                  books.title,
                  authors.author) AS rat
                  
              LEFT JOIN 
                  (SELECT
                       book_id,
                       COUNT(DISTINCT review_id) AS reviews_count
                    FROM
                        reviews
                    GROUP BY 
                       book_id) AS rev
               ON rev.book_id =rat.book_id
                            ;    
                         
          '''
pd.io.sql.read_sql(query_2, con = engine)

book_id                                              title  \
0          3  13 Little Blue Envelopes (Little Blue Envelope...   
1        821                                  The Pelican Brief   
2        836                                The Prince of Tides   
3        904                                     The Waste Land   
4        470          Neither Here nor There: Travels in Europe   
..       ...                                                ...   
995      972  Wherever You Go  There You Are: Mindfulness Me...   
996      929                                      Thunderstruck   
997      856                                   The Robber Bride   
998       46                                   A Scanner Darkly   
999      885  The Tipping Point: How Little Things Can Make ...   

                       author    rating  reviews_count  
0             Maureen Johnson  4.666667            3.0  
1                John Grisham  4.000000            3.0  
2                  Pat Conroy  4.142857            3.0  
3    T.S. Eliot/Michael North  3.500000            2.0  
4                 Bill Bryson  2.500000            2.0  
..                        ...       ...            ...  
995            Jon Kabat-Zinn  5.000000            2.0  
996               Erik Larson  2.500000            2.0  
997           Margaret Atwood  4.000000            2.0  
998            Philip K. Dick  3.666667            3.0  
999          Malcolm Gladwell  3.818182            5.0  

[1000 rows x 5 columns]

ВЫВОД:

В результате запроса выведены данные, где для каждой из 1000 книг приведены идентификатор книги,  наименование, автор,  количество обзоров и средняя оценка.

<a href = '#section1'>Вернуться к оглавлению</a>
</div>

<a class="ancor" id="step_5"></a>
## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [11]:
query_3 = ''' SELECT
                  books.publisher_id AS publisher_id,
                  publishers.publisher AS publisher,
                  COUNT (books.book_id) AS book_cnt
              FROM
                  books
                  INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
              WHERE 
                  books.num_pages>50
              GROUP BY
                  books.publisher_id,
                  publishers.publisher
              ORDER BY
                  book_cnt DESC
              LIMIT 1
                            ;    
                         
          '''
pd.io.sql.read_sql(query_3, con = engine)

publisher_id      publisher  book_cnt
0           212  Penguin Books        42

ВЫВОД:

Издательство Penguin Booksе выпустило наибольшее число книг, исключая брошюры, по сравнению с другими издательствами.

<a href = '#section1'>Вернуться к оглавлению</a>
</div>

<a class="ancor" id="step_6"></a>
## Определите автора с самой высокой оценкой книг - учитывайте только книги с 50 и более оценками

In [12]:
query_4 = ''' SELECT 
                  A.author AS author,
                  AVG(A.rating_avg) AS autor_avg
              FROM
             (SELECT
                  books.book_id AS book_id,
                  books.title AS title,
                  authors.author AS author,
                  AVG(ratings.rating) as rating_avg
              FROM
                  books
                  INNER JOIN authors ON authors.author_id = books.author_id
                  INNER JOIN ratings ON ratings.book_id = books.book_id
              GROUP BY
                  books.book_id,
                  books.title,
                  authors.author
              HAVING COUNT(ratings.rating_id) >= 50
              ORDER BY
                  rating_avg DESC) AS A
              GROUP BY
                    
                    A.author
                   
              ORDER BY
                    autor_avg DESC
              LIMIT 1;    
                         
          '''
pd.io.sql.read_sql(query_4, con = engine)

author  autor_avg
0  J.K. Rowling/Mary GrandPré   4.283844

ВЫВОД:

J.K. Rowling/Mary GrandPré -  автор с самой высокой оценкой книг, если учитывать только книги с 50 и более оценками.

<a href = '#section1'>Вернуться к оглавлению</a>
</div>

<a class="ancor" id="step_7"></a>
## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [14]:
query_5 = ''' SELECT 
                  AVG(R.cnt_rev) AS avg_rev
              FROM
             (SELECT
                  U.username AS username,
                  COUNT(reviews.review_id) AS cnt_rev
              FROM   
             (SELECT
                 username,
                 COUNT(rating_id) AS cnt
              FROM
                  ratings
              GROUP BY 
                  username                
              HAVING 
                  COUNT(rating_id) > 50) AS U
              INNER JOIN reviews ON  reviews.username = U.username
              GROUP BY 
                  U.username) AS R;    
                         
          '''
pd.io.sql.read_sql(query_5, con = engine)
        

avg_rev
0  24.333333

ВЫВОД:

В среднем пользователи, которые поставили больше 50 оценок, сделали 24 обзора.¶

<a href = '#section1'>Вернуться к оглавлению</a>
</div>